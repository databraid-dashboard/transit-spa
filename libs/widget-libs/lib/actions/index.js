import _regeneratorRuntime from 'babel-runtime/regenerator';

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return Promise.resolve(value).then(function (value) { step("next", value); }, function (err) { step("throw", err); }); } } return step("next"); }); }; }

import * as TYPES from '../constants/constants';

export function updateCurrentLocation(location) {
  return {
    type: TYPES.UPDATE_CURRENT_LOCATION,
    currentLocation: location
  };
}

export function addDestination(dest) {
  return {
    type: TYPES.ADD_DESTINATION,
    destination: dest
  };
}

export function addJourneys(destinationId, journeys) {
  return {
    type: TYPES.ADD_JOURNEYS,
    destinationId: destinationId,
    journeys: journeys
  };
}

export function removeJourneys(destinationId) {
  return {
    type: TYPES.REMOVE_JOURNEYS,
    destinationId: destinationId
  };
}

export function removeDestination(destinationId) {
  return {
    type: TYPES.REMOVE_DESTINATION,
    destinationId: destinationId
  };
}

<<<<<<< HEAD
export function refreshJourneys(destinationId, origin, destination) {
=======
export function fetchJourneys(destinationId, origin, destination) {
>>>>>>> Inital dashboard integration
  var _this = this;

  return function () {
    var _ref2 = _asyncToGenerator(_regeneratorRuntime.mark(function _callee(dispatch, getState, _ref) {
      var Api = _ref.Api;
<<<<<<< HEAD
      var json, journeys, journeysOffset;
=======
      var json, journeys;
>>>>>>> Inital dashboard integration
      return _regeneratorRuntime.wrap(function _callee$(_context) {
        while (1) {
          switch (_context.prev = _context.next) {
            case 0:
              _context.next = 2;
              return Api.fetchJourneys(origin, destination);

            case 2:
              json = _context.sent;
              journeys = json.map(function (rawJourneyObj) {
                var journeyObj = {
                  destination: rawJourneyObj.legs[0].end_address,
                  arrivalTimeText: rawJourneyObj.legs[0].arrival_time.text,
                  departureTimeUTC: rawJourneyObj.legs[0].departure_time.value,
                  transitSteps: rawJourneyObj.legs[0].steps.map(function (step) {
                    var stepObj = {
                      instruction: step.html_instructions,
                      mode: step.travel_mode,
                      duration: step.duration.text
                    };
                    return stepObj;
                  })
                };
                return journeyObj;
              });
<<<<<<< HEAD
              journeysOffset = journeys.filter(function (journey) {
                var currentTimeInSeconds = Date.now() / 1000;
                var diff = journey.departureTimeUTC - currentTimeInSeconds;
                var offset = 1.5;
                return diff >= offset;
              }).sort(function (a, b) {
                return a - b;
              });


              dispatch(removeJourneys(destinationId));
              dispatch(addJourneys(destinationId, journeysOffset));

            case 7:
=======
              return _context.abrupt('return', dispatch(addJourneys(destinationId, journeys)));

            case 5:
>>>>>>> Inital dashboard integration
            case 'end':
              return _context.stop();
          }
        }
      }, _callee, _this);
    }));

    return function (_x, _x2, _x3) {
      return _ref2.apply(this, arguments);
    };
  }();
<<<<<<< HEAD
}

export function fetchJourneys(destinationId, origin, destination) {
  var _this2 = this;

  return function () {
    var _ref4 = _asyncToGenerator(_regeneratorRuntime.mark(function _callee2(dispatch, getState, _ref3) {
      var Api = _ref3.Api;
      var json, journeys;
      return _regeneratorRuntime.wrap(function _callee2$(_context2) {
        while (1) {
          switch (_context2.prev = _context2.next) {
            case 0:
              _context2.next = 2;
              return Api.fetchJourneys(origin, destination);

            case 2:
              json = _context2.sent;
              journeys = json.map(function (rawJourneyObj) {
                var journeyObj = {
                  destination: rawJourneyObj.legs[0].end_address,
                  arrivalTimeText: rawJourneyObj.legs[0].arrival_time.text,
                  departureTimeUTC: rawJourneyObj.legs[0].departure_time.value,
                  transitSteps: rawJourneyObj.legs[0].steps.map(function (step) {
                    var stepObj = {
                      instruction: step.html_instructions,
                      mode: step.travel_mode,
                      duration: step.duration.text,
                      shortName: step.transit_details ? step.transit_details.line.short_name : '',
                      agency: step.transit_details ? step.transit_details.line.agencies[0].name : ''
                    };
                    return stepObj;
                  })
                };
                return journeyObj;
              });


              dispatch(addJourneys(destinationId, journeys));

            case 5:
            case 'end':
              return _context2.stop();
          }
        }
      }, _callee2, _this2);
    }));

    return function (_x4, _x5, _x6) {
      return _ref4.apply(this, arguments);
    };
  }();
=======
>>>>>>> Inital dashboard integration
}