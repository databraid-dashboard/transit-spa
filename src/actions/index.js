import * as TYPES from '../constants/constants';
import * as HELPERS from '../utils/helpers';

export function updateCurrentLocation() {
  return async (dispatch, getState, { TRANSIT_API }) => {
    dispatch({
      type: TYPES.GEOLOCATING,
    });

    const location = await TRANSIT_API.getCurrentLocation();

    return dispatch({
      type: TYPES.UPDATE_CURRENT_LOCATION,
      currentLocation: location,
    });
  };
}

export function addDestination(dest) {
  return {
    type: TYPES.ADD_DESTINATION,
    destination: dest,
  };
}

export function addJourneys(destinationId, journeys) {
  return {
    type: TYPES.ADD_JOURNEYS,
    destinationId,
    journeys,
  };
}

export function removeJourneys(destinationId) {
  return {
    type: TYPES.REMOVE_JOURNEYS,
    destinationId,
  };
}

export function removeDestination(destinationId) {
  return {
    type: TYPES.REMOVE_DESTINATION,
    destinationId,
  };
}

export function fetchJourneys(destinationId, origin, destination) {
  return async (dispatch, getState, { TRANSIT_API }) => {
    // NOTE: json also includes fare{}, bounds{}, and polyline{}--not being used
    const json = await TRANSIT_API.fetchJourneys(origin, destination);
    const alerts = await TRANSIT_API.fetchAlerts();

    const journeys = await HELPERS.createArrayOfJourneyObjects(json);
    const journeysWithAlerts = await HELPERS.applyAlerts(journeys, alerts);
    const journeysOffset = await HELPERS.offsetJourneys(journeysWithAlerts);

    const currentJourneysForDestination = getState().widgets.byId.transit.journeys.byDestinationId[
      destinationId
    ];

    if (currentJourneysForDestination) {
      dispatch(removeJourneys(destinationId));
    }

    dispatch(addJourneys(destinationId, journeysOffset));
    dispatch({
      type: TYPES.ALERTS_RETRIEVED,
      alerts,
    });
  };
}
