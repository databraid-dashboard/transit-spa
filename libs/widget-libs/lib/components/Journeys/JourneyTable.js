var _jsxFileName = '../../src/components/Journeys/JourneyTable.jsx';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import BestJourney from './BestJourney';
import NextBestJourney from './NextBestJourney';

import { fetchJourneys } from '../../actions';

export function timeToLeaveConverter(departureTimeInSeconds) {
  var currentTimeInSeconds = Date.now() / 1000;
  var diff = departureTimeInSeconds - currentTimeInSeconds;

  return Math.floor(diff);
}

export var JourneyTable = function (_Component) {
  _inherits(JourneyTable, _Component);

  function JourneyTable() {
    _classCallCheck(this, JourneyTable);

    return _possibleConstructorReturn(this, (JourneyTable.__proto__ || Object.getPrototypeOf(JourneyTable)).apply(this, arguments));
  }

  _createClass(JourneyTable, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      var _props = this.props,
          destinationId = _props.destinationId,
          origin = _props.origin,
          destinationsById = _props.destinationsById;

      this.props.fetchJourneys(destinationId, origin, destinationsById[destinationId].address);
    }
  }, {
    key: 'render',
    value: function render() {
      var journeys = this.props.journeys;


      if (!journeys) return React.createElement(
        'div',
        {
          __source: {
            fileName: _jsxFileName,
            lineNumber: 28
          },
          __self: this
        },
        'Loading...'
      );

      var bestJourney = journeys[0];
      var nextBestJourney = journeys[1];

      var timeToLeaveBest = timeToLeaveConverter(bestJourney.departureTimeUTC);
      var timeToLeaveNextBest = timeToLeaveConverter(nextBestJourney.departureTimeUTC);

      return React.createElement(
        'div',
        {
          __source: {
            fileName: _jsxFileName,
            lineNumber: 37
          },
          __self: this
        },
        React.createElement(BestJourney, {
          timeToLeaveInSeconds: timeToLeaveBest,
          steps: bestJourney.transitSteps,
          eta: bestJourney.arrivalTimeText,
          conditionStatus: 'on-time',
          __source: {
            fileName: _jsxFileName,
            lineNumber: 38
          },
          __self: this
        }),
        React.createElement(NextBestJourney, {
          timeToLeaveInSeconds: timeToLeaveNextBest,
          steps: nextBestJourney.transitSteps,
          eta: nextBestJourney.arrivalTimeText,
          conditionStatus: 'future undertain -- see journey table',
          __source: {
            fileName: _jsxFileName,
            lineNumber: 44
          },
          __self: this
        })
      );
    }
  }]);

  return JourneyTable;
}(Component);

JourneyTable.propTypes = {
  destinationId: PropTypes.number.isRequired,
  origin: PropTypes.string.isRequired,
  destinationsById: PropTypes.shape({
    1: PropTypes.object
  }).isRequired,
  fetchJourneys: PropTypes.func.isRequired,
  journeys: PropTypes.arrayOf(PropTypes.object)
};

JourneyTable.defaultProps = {
  destinationId: 1,
  origin: '',
  destinationsById: { 1: {} },
  fetchJourneys: function fetchJourneys() {},
  journeys: [{
    departureTimeUTC: Date.now(),
    arrivalTimeText: '00:00am',
    transitSteps: [{ duration: '1 mins', instruction: 'Walk to Some St. Station', mode: 'WALKING' }, {
      duration: '23 mins',
      instruction: 'Metro rail towards Warm Springs/South Fremont',
      mode: 'TRANSIT'
    }]
  }, {
    departureTimeUTC: Date.now(),
    arrivalTimeText: '00:00pm',
    transitSteps: [{ duration: '4 mins', instruction: 'Walk to Some St. Station', mode: 'WALKING' }, {
      duration: '56 mins',
      instruction: 'Metro rail towards Warm Springs/South Fremont',
      mode: 'TRANSIT'
    }]
  }]
};
export var mapStateToProps = function mapStateToProps(state, ownProps) {
  var origin = state.configuration.currentLocation.address;
  var destinationsById = state.destinations.byId;
  var destinationId = ownProps.id;
  var journeys = state.journeys.byDestinationId[destinationId];

  return {
    origin: origin,
    journeys: journeys,
    destinationId: destinationId,
    destinationsById: destinationsById
  };
};

export var mapDispatchToProps = function mapDispatchToProps(dispatch) {
  return bindActionCreators({
    fetchJourneys: fetchJourneys
  }, dispatch);
};

export default connect(mapStateToProps, mapDispatchToProps)(JourneyTable);