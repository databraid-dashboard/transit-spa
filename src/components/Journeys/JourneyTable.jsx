import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { fetchJourneys, refreshJourneys } from '../../actions';

import BestJourney from './BestJourney';
import NextBestJourney from './NextBestJourney';

import injectWidgetId from '../../utils/utils';

export function timeToLeaveConverter(departureTimeInSeconds) {
  const currentTimeInSeconds = Date.now() / 1000;
  const diff = departureTimeInSeconds - currentTimeInSeconds;
  return Math.floor(diff);
}

export class JourneyTable extends Component {
  constructor(props) {
    super(props);

    this.callRefreshJourneys = this.callRefreshJourneys.bind(this);
  }

  componentDidMount() {
    const { destinationId, origin, destinationsById } = this.props;
    this.props.fetchJourneys(destinationId, origin, destinationsById[destinationId].address);
  }

  callRefreshJourneys() {
    const { destinationId, origin, destinationsById } = this.props;
    this.props.refreshJourneys(destinationId, origin, destinationsById[destinationId].address);
  }

  render() {
    const { journeys } = this.props;

    if (!journeys || journeys.length === 0) return <div>Loading...</div>;

    const bestJourney = journeys[0];
    const nextBestJourney = journeys[1];

    const timeToLeaveBest = timeToLeaveConverter(bestJourney.departureTimeUTC);
    const timeToLeaveNextBest = timeToLeaveConverter(nextBestJourney.departureTimeUTC);

    return (
      <div>
        <BestJourney
          timeToLeaveInSeconds={timeToLeaveBest}
          steps={bestJourney.transitSteps}
          eta={bestJourney.arrivalTimeText}
          conditionStatus={'on-time'}
          callRefreshJourneys={this.callRefreshJourneys}
        />
        <NextBestJourney
          timeToLeaveInSeconds={timeToLeaveNextBest}
          steps={nextBestJourney.transitSteps}
          eta={nextBestJourney.arrivalTimeText}
          conditionStatus={'future undertain -- see journey table'}
          callRefreshJourneys={this.callRefreshJourneys}
        />
      </div>
    );
  }
}

JourneyTable.propTypes = {
  destinationId: PropTypes.number.isRequired,
  origin: PropTypes.string.isRequired,
  destinationsById: PropTypes.shape({
    1: PropTypes.object,
  }).isRequired,
  fetchJourneys: PropTypes.func.isRequired,
  refreshJourneys: PropTypes.func.isRequired,
  journeys: PropTypes.arrayOf(PropTypes.object),
};

JourneyTable.defaultProps = {
  destinationId: 1,
  origin: '',
  destinationsById: { 1: {} },
  fetchJourneys: () => {},
  refreshJourneys: () => {},
  journeys: [
    {
      departureTimeUTC: Date.now(),
      arrivalTimeText: '00:00am',
      transitSteps: [
        { duration: '1 mins', instruction: 'Walk to Some St. Station', mode: 'WALKING' },
        {
          duration: '23 mins',
          instruction: 'Metro rail towards Warm Springs/South Fremont',
          mode: 'TRANSIT',
        },
      ],
    },
    {
      departureTimeUTC: Date.now(),
      arrivalTimeText: '00:00pm',
      transitSteps: [
        { duration: '4 mins', instruction: 'Walk to Some St. Station', mode: 'WALKING' },
        {
          duration: '56 mins',
          instruction: 'Metro rail towards Warm Springs/South Fremont',
          mode: 'TRANSIT',
        },
      ],
    },
  ],
};
export const mapStateToProps = (state, ownProps) => {
  const id = ownProps.widgetId;
  const origin = state.widgets.byId[id].configuration.currentLocation.address;
  const destinationsById = state.widgets.byId[id].destinations.byId;
  const destinationId = ownProps.id;
  const journeys = state.widgets.byId[id].journeys.byDestinationId[destinationId];

  return {
    origin,
    journeys,
    destinationId,
    destinationsById,
  };
};

export const mapDispatchToProps = dispatch =>
  bindActionCreators(
    {
      fetchJourneys,
      refreshJourneys,
    },
    dispatch,
  );

export default injectWidgetId(connect(mapStateToProps, mapDispatchToProps)(JourneyTable));
