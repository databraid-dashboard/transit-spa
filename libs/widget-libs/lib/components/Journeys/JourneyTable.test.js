var _jsxFileName = '../../src/components/Journeys/JourneyTable.test.jsx',
    _this = this;

import React from 'react';
import { shallow, render, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { JourneyTable, timeToLeaveConverter, mapStateToProps, mapDispatchToProps } from './JourneyTable';
import { fetchJourneys } from '../../actions';

describe('JourneyTable', function () {
  var testData = {
    destinationId: 1,
    origin: '',
    destinationsById: { 1: {} },
    journeys: [{
      departureTimeUTC: 1502222222,
      arrivalTimeText: '00:00am',
      transitSteps: [{ duration: '1 mins', instruction: 'Walk to Some St. Station', mode: 'WALKING' }, {
        duration: '23 mins',
        instruction: 'Metro rail towards Warm Springs/South Fremont',
        mode: 'TRANSIT'
      }]
    }, {
      departureTimeUTC: 1502222555,
      arrivalTimeText: '00:00pm',
      transitSteps: [{ duration: '4 mins', instruction: 'Walk to Some St. Station', mode: 'WALKING' }, {
        duration: '56 mins',
        instruction: 'Metro rail towards Warm Springs/South Fremont',
        mode: 'TRANSIT'
      }]
    }],
    timeToLeaveInSeconds: 1502222223
  };

  it('should render Loading...', function () {
    var component = shallow(React.createElement(JourneyTable, {
      id: 1,
      destinationId: testData.destinationId,
      origin: testData.origin,
      destinationsById: testData.destinationsById,
      journeys: null,
      __source: {
        fileName: _jsxFileName,
        lineNumber: 48
      },
      __self: _this
    }));
    expect(toJson(component)).toMatchSnapshot();
  });

  it('mapDispatchToProps', function () {
    var dispatch = jest.fn();
    var fetchJourneys = jest.fn();
    var bindActionCreators = jest.fn();

    expect(mapDispatchToProps(dispatch)).toHaveProperty('fetchJourneys');
  });

  it('calls fetchJourneys on load', function () {
    var fetchJourneys = jest.fn();
    var component = mount(React.createElement(JourneyTable, {
      id: 1,
      destinationId: testData.destinationId,
      origin: testData.origin,
      destinationsById: testData.destinationsById,
      journeys: testData.journeys,
      fetchJourneys: fetchJourneys,
      __source: {
        fileName: _jsxFileName,
        lineNumber: 70
      },
      __self: _this
    }));

    expect(fetchJourneys).toHaveBeenCalled();
  });
});

describe('timeToLeaveConverter', function () {
  it('should return positive integer', function () {
    var futureTimeInSeconds = Date.now() + 1000;
    expect(timeToLeaveConverter(futureTimeInSeconds)).toBeGreaterThanOrEqual(1);
  });
});