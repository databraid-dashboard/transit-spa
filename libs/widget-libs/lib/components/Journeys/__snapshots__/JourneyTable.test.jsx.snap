// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JourneyTable should render Loading... 1`] = `
<div>
  Loading...
</div>
`;
